stages/stage_2/optimize.py

```diff
@@ -193,19 +198,47 @@ def path_optimizer(r, eval):
     """
 
     t1 = match(FIRST_ATOM_PATTERN, r)
-    if t1:
+    if t1 and non_nil(t1["atom"]):
         node = NodePath(t1["atom"].as_int())
         node = node + LEFT
         return r.to(node.as_short_path())
 
     t1 = match(REST_ATOM_PATTERN, r)
-    if t1:
+    if t1 and non_nil(t1["atom"]):
         node = NodePath(t1["atom"].as_int())
         node = node + RIGHT
         return r.to(node.as_short_path())
     return r
 
 
+QUOTE_PATTERN_1 = assemble("(q . 0)")
+
+
+def quote_null_optimizer(r, eval):
```

Comment:
This one I'm a little worried about.

In the context of compiling a program, I think it's 100% good.
But if the ChiaLisp compiler produces output that is then used as an environment for another program, it introduces a non-orthogonal transform for other values.

eg. if the consuming program expects a list of quoted ints.

Reply:
Right, you should definitely not run the optimizer on anything other than code you expect to run (and not analyze at runtime).